File: incre-tests/synduce/ptree/maxlast.f
Inductive Tree = telt Int | tnode {Int, Tree, Tree};
Inductive PTree = pelt Int | pnode {Int, List}
    with List = elt PTree | cons {PTree, List};

max = \a: Int. \b: Int. if < a b then b else a;

last = fix (
  \f: Tree -> Int. \t: Tree.
  match t with
    telt w -> w
  | tnode {_, _, r} -> f r
  end
);

repr = fix (
  \f: Tree -> PTree. \x: Tree.
  match x with
    telt a -> pelt a
  | tnode {a, l, r} -> 
    let repr_list = fix (
        \g: Tree -> List. \y: Tree.
        match y with
          telt a -> let z = pelt a in elt z
        | tnode {a, l, r} -> cons {pelt a, cons {f l, g r}}
        end
    ) in
    pnode {a, cons {f l, repr_list r}}
  end
);

spec = fix (
  \f: PTree -> Int. \x: PTree.
  match x with
    pelt a -> a
  | pnode {a, l} -> 
    let maxh_aux = fix (
        \g: List -> Int. \y: List.
        match y with
          elt a -> f a
        | cons {hd, tl} -> g tl
        end
    ) in 
    max a (maxh_aux l)
  end
);

target = fix (
  \f: Tree -> Compress Tree. \t: Tree.
  match t with
    telt w -> t
  | tnode {w, l, r} -> tnode {w, f l, f r}
  end
);

main = \t: Tree. spec (repr (target t));

File: ml-benchmark/synduce/ptree/maxlast.ml
type tree = Telt of int | Tnode of int * tree * tree

type ptree = Pelt of int | Pnode of int * plist
and plist = Elt of ptree | Cons of ptree * plist

let max a b = if a < b then b else a

let rec last t =
  match t with
  | Telt w -> w
  | Tnode (_, _, r) -> last r

let rec repr x =
  match x with
  | Telt a -> Pelt a
  | Tnode (a, l, r) ->
      let rec repr_list y =
        match y with
        | Telt a ->
            let z = Pelt a in
            Elt z
        | Tnode (a, l, r) ->
            Cons (Pelt a, Cons (repr l, repr_list r))
      in
      Pnode (a, Cons (repr l, repr_list r))

let rec spec x =
  match x with
  | Pelt a -> a
  | Pnode (a, l) ->
      let rec maxh_aux y =
        match y with
        | Elt a -> spec a
        | Cons (hd, tl) -> maxh_aux tl
      in
      max a (maxh_aux l)

let rec target t =
  match t with
  | Telt w -> t
  | Tnode (w, l, r) -> Tnode (w, target l, target r)

let program t = spec (repr (target t))